File: Algorithms/1268_Search_Suggestions_System/Python/Search_Suggestions_System_Solution_1.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def suggestedProducts(self, products, searchWord):
        cache = collections.defaultdict(list)
        res = []
        for word in products:
            for i in range(1, len(word) + 1):
                cache[word[:i]].append(word)

        for i in range(1, len(searchWord) + 1):
            if searchWord[:i] in cache:
                res.append(sorted(cache[searchWord[:i]])[:3])
            else:
                res.append([])

        return res

File: Algorithms/1268_Search_Suggestions_System/Java/SearchSuggestionsSystemSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Map<String, List<String>> cache = new HashMap<>();
        List<List<String>> res = new ArrayList<>();
        for (String word : products) {
            for (int i = 1; i <= word.length(); i++) {
                if (cache.containsKey(word.substring(0, i))) cache.get(word.substring(0, i)).add(word);
                else cache.put(word.substring(0, i), new ArrayList<>(Arrays.asList(word)));
            }
        }
        for (int i = 1; i <= searchWord.length(); i++) {
            List<String> temp = cache.getOrDefault(searchWord.substring(0, i), new ArrayList<>());
            if (temp.size() != 0) {
                temp.sort(String::compareTo);
                if (temp.size() > 3) temp = temp.subList(0, 3);
            }
            res.add(temp);
        }
        return res;
    }
}





